### DISCUSSION TITLE
fix "BlInitializeLibrary failed 0xc000009a"

### DISCUSSION THREAD

**Discussion by acoul** [a1ive/grub#100] [Q&A] [answered]
Greetings,

I am looking at the [add fixmmap command to fix "BlInitializeLibrary failed 0xc000009a" Error](https://github.com/a1ive/grub/commit/52f551cb2cced3213be3a94c0158565e2dc6e921) commit and wondering if this can become a .efi script so to be directly included in a UEFI menu on a FAT16/32 EFI SSD partition

**Comment by a1ive** [comment]
> if this can become a .efi script so to be directly included in a UEFI menu

What bootloader are you using currently?

**Reply by acoul** [reply]
> 
> What bootloader are you using currently?

I am using windows 7 & 8 on an msdos/ntfs partitioned disk along with [uefi-ntfs](https://github.com/pbatard/uefi-ntfs) for both mbr & uefi boot. on that second EFI/FAT12 partition there is an EFI\Boot directory that contains bootx64.efi & on that same directory I would be interested to place the BlInitializeLibrary.efi script in-order to call it directly from the UEFI boot menu

I am using [BootIce](https://www.majorgeeks.com/files/details/bootice_64_bit.html) for creating/editing UEFI menu entries, as I am not too familiar yet with the terminal commands

on that same directory, I have successfully installed [grubfmx64.efi](https://github.com/a1ive/grub2-filemanager) along with [EfiGuard Loader.efi](https://github.com/Mattiwatti/EfiGuard) & both work wonderfully

**Reply by a1ive** [reply]
I mean how you boot Windows

**Reply by acoul** [reply]
> I mean how you boot Windows

from the Bios UEFI menu (F12 for Gigabyte motherboards)

![win8_msdos_ntfs_uefi_class_3](https://github.com/a1ive/grub/assets/778451/f8ad6018-a46f-48c2-920a-b3b739c582f5)

![win8_uefi_minitool](https://github.com/a1ive/grub/assets/778451/b1243319-82a6-4838-98e4-95ba91f6efe0)


the contents of the EFI\Boot partition:

![EFI_Boot_content](https://github.com/a1ive/grub/assets/778451/341dd6ad-45b4-4629-9898-b9e294ac3d7f)

**Reply by a1ive** [reply]
What's the content of boot config menu? The 'bootx64.efi' is bootmgfw.efi or something else?

**Reply by acoul** [reply]
> What's the content of boot config menu? The 'bootx64.efi' is bootmgfw.efi or something else?

my bad, mainly I use Windows Boot Manager & BCD. sometimes, I also use the alternative bootx64.efi method. Either method (bootx64.efi or bootmgfw.efi ) as a solution would serve my needs fine

**Reply by acoul** [reply]
> > if this can become a .efi script so to be directly included in a UEFI menu
> 
> What bootloader are you using currently?

the original request is about the possibility of embedding the relative [BlInitializeLibrary fix code](https://github.com/a1ive/grub/commit/52f551cb2cced3213be3a94c0158565e2dc6e921) as a .efi script that can be run directly from the bios [UEFI Boot Manager](https://oofhours.com/2022/06/29/geeking-out-with-the-uefi-boot-manager/)

the bootloader should be irrelevant to the .efi script that I am looking for, or am I dead wrong on this ?

**Edit**: are you looking for something like this perhaps:
`bcdedit /store v:\EFI\Microsoft\Boot\BCD`
where v: is the FAT32 EFI partition

![win8_uefi_bcdedit_enum](https://github.com/a1ive/grub/assets/778451/bb0c5aca-ee1d-44dd-ac6c-952ebd34e32d)

**Reply by a1ive** [reply]
`.efi` files are EFI applications or EFI drivers, they are not scripts.
If you are booting Windows Manager directly from the firmware's UEFI boot menu, it should not be possible to get the error "BlInitializeLibrary failed XXX".

**Reply by a1ive** [reply]
All motherboards are tested for Windows compatibility before they are shipped.
If it can't boot Windows, wouldn't that be ridiculous and who would buy it?

**Reply by a1ive** [reply]
Only if the user is using a third-party boot manager (e.g. GRUB 2, rEFInd, etc.) may this issue need to be addressed.

**Reply by acoul** [reply]
> Only if the user is using a third-party boot manager (e.g. GRUB 2, rEFInd, etc.) may this issue need to be addressed.

on my [GA-A320M-H](https://www.gigabyte.com/Motherboard/GA-A320M-H-rev-1x#kf) motherboard (which supports CSM), if I press F12 in-order to get to the boot menu selection & I choose either Windows Boot Manager or UEFI OS, then I get the BlInitializeLibrary failed 0xc000009a error.

pre-Setting on the BIOS the desired boot device & then booting (without calling the UEFI boot menu) works fine without the BlInitializeLibrary error.

So far, I use [Ventoy's fix](https://forums.ventoy.net/showthread.php?tid=1274&pid=2448#pid2448) in combination with [grubfm.iso](https://github.com/a1ive/grub2-filemanager) to overcome this issue, but it would be handy if this code could be embedded as an .efi driver & included as a menu/choice on the Bios UEFI boot menu on the local disk

**Reply by a1ive** [reply, marked as answer]
In grubfm [F4] menu:
![1](https://github.com/a1ive/grub/assets/10670106/17b7a667-f926-45de-846f-faefaf1b8ead)
If you want to get a standalone efi to solve this problem, then you can build grub efi yourself (add fixmmap.mod), search and boot Windows using the following menu:
```
fixmmap
search -s -f /EFI/Microsoft/Boot/bootmgfw.efi
chainloader /EFI/Microsoft/Boot/bootmgfw.efi
```

**Comment by acoul** [comment]
[grubfm](https://github.com/a1ive/grub2-filemanager) is next best thing to morning coffee. a must for every bootloader. I was unaware of that BlInitializeLibrary fix menu which resolves this issue

thank you
